java-generic-mqtt/src/main/java/com/cumulocity/generic/mqtt/client/websocket/GenericMqttWebSocketPublisher.java

```diff
@@ -0,0 +1,84 @@
+import com.cumulocity.sdk.client.messaging.notifications.TokenApi;
+import org.apache.commons.codec.binary.Base64;
+
+import java.net.URI;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.apache.commons.lang3.RandomStringUtils.randomAlphabetic;
+
+class GenericMqttWebSocketPublisher implements GenericMqttPublisher {
+
+    private final static String WEBSOCKET_URL_PATTERN = "%s/notification2/producer/?token=%s";
+    private static final String TOKEN_TYPE = "mqtt";
+
+    private final String webSocketBaseUrl;
+    private final TokenApi tokenApi;
+
+    private final GenericMqttMessageConverter genericMqttMessageConverter = new GenericMqttMessageConverter();
+
+    private final AtomicInteger sequence = new AtomicInteger();
+    private final GenericMqttConnectionConfig config;
+
+    private GenericMqttWebSocketClient producer;
+
+    public GenericMqttWebSocketPublisher(String webSocketBaseUrl, TokenApi tokenApi, GenericMqttConnectionConfig config) {
+        this.webSocketBaseUrl = webSocketBaseUrl;
+        this.tokenApi = tokenApi;
+        this.config = config;
+    }
+
+    @Override
+    public void publish(GenericMqttMessage genericMqttMessage) {
+        final NotificationTokenRequestRepresentation tokenRequestRepresentation = getTokenRepresentation(config.getTopic());
+        final String token = tokenApi
+                .create(tokenRequestRepresentation)
+                .getTokenString();
+
+        if (token == null) {
+            throw new GenericMqttWebSocketClientException(String.format("Token could not be created for topic %s", config.getTopic()));
+        }
+
+        if (producer == null) {
+            try {
+                final URI uri = new URI(String.format(WEBSOCKET_URL_PATTERN, webSocketBaseUrl, token));
+                producer = new GenericMqttWebSocketClient(uri);
+                producer.connectBlocking(config.getConnectionTimeoutInMillis(), TimeUnit.MILLISECONDS);
+            } catch (Exception e) {
+                throw new GenericMqttWebSocketClientException("Error publishing message!", e);
+            }
+        }
+
+        final byte[] data = genericMqttMessageConverter.encode(genericMqttMessage);
+        final String publishMessage = sequence.incrementAndGet() + "\n" + Base64.encodeBase64String(data);
+
+        producer.send(publishMessage);
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (producer != null) {
+            producer.close();
+        }
+    }
+
+    private NotificationTokenRequestRepresentation getTokenRepresentation(String topic) {
+        final NotificationTokenRequestRepresentation representation = new NotificationTokenRequestRepresentation();
+        representation.setSubscriber("javaGenericMqttWSClientPub" + randomAlphabetic(10));
+        representation.setSubscription(topic);
+        representation.setType(TOKEN_TYPE);
+        representation.setSigned(true);
```

Comment:
nitpick: 1440 is a magic number. Maybe a constant TOKEN_DURATION_MINUTES = 1440.
But ...
issue: we should document (somewhere in the creation methods) that the publishers/subscribers are created with Tokens that expire after 1 day (1440 minutes).

Reply:
What happens if token will expire? Client will receive exception and it has to close current publisher/subscriber and create new one?

Reply:
In the session is not closed, then no impact. If new producer, new token then.